### PR TITLE
Zerostudents

### LINKED ISSUE #326: [BUG] [Projects] zero students projects
now you can make a project with 0 students

### PR DESCRIPTION
**Pull request by WardM99**
close #326

### CODE REVIEW

File: backend/src/app/exceptions/handlers.py [hunk 1]
```diff
@@ -140,3 +140,10 @@ def failed_to_add_new_email_exception(_request: Request, _exception: FailedToAdd
             status_code=status.HTTP_400_BAD_REQUEST,
             content={'message': 'Something went wrong while creating a new email'}
         )
+
+    @app.exception_handler(NoneStrictPostiveNumberOfSlots)
+    def none_strict_postive_number_of_slots(_request: Request, _exception: NoneStrictPostiveNumberOfSlots):
+        return JSONResponse(
+            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
+            content={'message': 'Roles has to be strict positive'}
```
Comment (stijndcl) [inline backend/src/app/exceptions/handlers.py:148]:
```suggestion
            content={'message': 'The amount of slots per role has to be a strictly positive integer'}
```

File: backend/src/app/exceptions/projects.py [hunk 2]
```diff
@@ -8,3 +8,7 @@ class FailedToAddProjectRoleException(Exception):
     """
     Exception raised when a project_role can't be added for some reason
     """
+
+
+class NoneStrictPostiveNumberOfSlots(Exception):
+    """Exceptioon raised when roles aren't strict postive"""
```
Comment (stijndcl) [inline backend/src/app/exceptions/projects.py:14]:
```suggestion
    """Exception raised when roles aren't strictly positive"""
```

File: backend/src/app/routers/editions/projects/projects.py [hunk 3]
```diff
@@ -115,15 +116,15 @@ async def post_project_role(
 
 @projects_router.patch(
     "/{project_id}/roles/{project_role_id}",
-    response_model=ProjectRoleSchema,
+    status_code=status.HTTP_204_NO_CONTENT,
```
Comment (stijndcl) [inline backend/src/app/routers/editions/projects/projects.py:119]: Any reason why you're removing the return value here?

File: backend/src/app/exceptions/projects.py [hunk 4]
```diff
@@ -8,3 +8,7 @@ class FailedToAddProjectRoleException(Exception):
     """
     Exception raised when a project_role can't be added for some reason
     """
+
+
+class NoneStrictPostiveNumberOfSlots(Exception):
```
Comment (stijndcl) [inline backend/src/app/exceptions/projects.py:13]: (This should also be changed everywhere where the exception is used, not just in here)
```suggestion
class NoStrictlyPositiveNumberOfSlots(Exception):
```

File: backend/tests/test_routers/test_editions/test_projects/test_projects.py [hunk 5]
```diff
@@ -332,6 +334,242 @@ async def test_delete_project_role(database_session: AsyncSession, auth_client:
...
+            "description": "description",
+            "slots": 1
+        })
+        assert response.status_code == status.HTTP_201_CREATED
+        response = await auth_client.patch("/editions/ed2022/projects/1/roles/1", json={
+            "skill_id": 1,
+            "description": "description",
+            "slots": 2
+        })
+        assert response.status_code == status.HTTP_204_NO_CONTENT
```
Comment (stijndcl) [inline backend/tests/test_routers/test_editions/test_projects/test_projects.py:468]: You should probably check if the change actually went through here, not just what the status code is

**Review by stijndcl — CHANGES_REQUESTED** [review]
Some minor remarks. Good to see tests for project roles, coverage++

File: backend/src/app/routers/editions/projects/projects.py (hunk 3 above)
Comment (WardM99) [reply backend/src/app/routers/editions/projects/projects.py:120]: @stijndcl because otherwise I get this error with pytest 
image

Comment (stijndcl) [reply backend/src/app/routers/editions/projects/projects.py:119]: Response is supposed to be response_class in the decorator, not response_model. If you rename that the error is gone. That's no reason to change this.

Comment (stijndcl) [reply backend/src/app/routers/editions/projects/projects.py:119]: Then just change the status code to something else than 204.

File: backend/src/app/logic/projects.py [hunk 6]
```diff
@@ -67,13 +68,17 @@ async def get_project_roles(db: AsyncSession, project: Project) -> ProjectRoleRe
 
 async def create_project_role(db: AsyncSession, project: Project, input_project_role: InputProjectRole) -> ProjectRole:
     """Create a project role"""
-    return await crud.create_project_role(db, project, input_project_role)
+    if(input_project_role.slots) > 0:
```
Comment (FKD13) [inline backend/src/app/logic/projects.py:71]:
```suggestion
    if input_project_role.slots > 0:
```

File: backend/src/app/logic/projects.py [hunk 7]
```diff
@@ -67,13 +68,17 @@ async def get_project_roles(db: AsyncSession, project: Project) -> ProjectRoleRe
...
+        return await crud.create_project_role(db, project, input_project_role)
+    raise NoStrictlyPositiveNumberOfSlots
 
 
 async def patch_project_role(db: AsyncSession, project_role_id: int, input_project_role: InputProjectRole) \
         -> ProjectRole:
     """Update a project role"""
-    return await crud.patch_project_role(db, project_role_id, input_project_role)
+    if(input_project_role.slots) > 0:
```
Comment (FKD13) [inline backend/src/app/logic/projects.py:79]:
```suggestion
    if input_project_role.slots > 0:
```

File: backend/tests/test_routers/test_editions/test_projects/test_projects.py [hunk 8]
```diff
@@ -24,7 +24,8 @@ async def test_get_projects_paginated(database_session: AsyncSession, auth_clien
         assert len(response.json()['projects']) == DB_PAGE_SIZE
         response = await auth_client.get("/editions/ed2022/projects?page=1", follow_redirects=True)
         assert response.status_code == status.HTTP_200_OK
-        assert len(response.json()['projects']) == round(DB_PAGE_SIZE * 1.5) - DB_PAGE_SIZE
+        assert len(response.json()['projects']) == round(
+            DB_PAGE_SIZE * 1.5) - DB_PAGE_SIZE
```
Comment (FKD13) [inline backend/tests/test_routers/test_editions/test_projects/test_projects.py:28]: This is shorter than the 120 character limit

**Review by FKD13 — CHANGES_REQUESTED** [review]

**Review by FKD13 — APPROVED** [review]

**Review by stijndcl — APPROVED** [review]

**Pull request merged**